### PR TITLE
Upgrade to weaviate-client to v4

### PR DESCRIPTION
**Pull request by Lee-W**
## Why
According to [Weaviate Python client (v4) goes GA](https://weaviate.io/blog/py-client-v4-release), the v3 client is now deprecated and we need to upgrade to v4

## What
* update min weaviate-client version to 4.4.0
* weaviate hook
    * update airflow connection to v4 style
    * migrate the following methods to v4 API
        * `test_connections`
        * `create_class` to `create_collection`
        * `get_schema` to `get_collection_configuraiton`
        * `delete_classes` to `delete_collections`
        * `query_with_vector`
        * `query_without_vector` to `query_with_text`
        * `create_object`
        * `get_object`
        * `delete_object`
        * `update_object`
        * `replace_object`
        * `object_exists`
        * `batch_data`
        * `get_or_create_object`
        * `update_collection_configuration`
        * `_generate_uuids`
        * `create_or_replace_document_objects`
        * `_delete_objects`
        * `_delete_all_documents_objects`
        * `_get_documents_to_uuid_map`
        * `_get_segregated_documents`
    * remove the following methods which are no longer supported in v4 API
        * `validate_object`
        * `update_schema`
        * `create_schema`
        * `delete_all_schema`
        * `create_or_replace_classes`
        * `_compare_schema_subset`
        * `_convert_properties_to_dict`
        * `check_subset_of_schema`
    * remove deprecated method `get_client`
    * remove unused argument retry_status_codes in `__init__`
* weaviate operators
    * migrate `WeaviateDocumentIngestOperator` and `WeaviateIngestOperator` to use hook with v4 API

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by Lee-W** [comment]
Hi @eladkal , if my memory serves me right, when we make a PR with breaking changes to a provider, we need to mark it somewhere. Could you please point out where I should mark it? Thanks!

**Comment by potiuk** [comment]
> Hi @eladkal , if my memory serves me right, when we make a PR with breaking changes to a provider, we need to mark it somewhere. Could you please point out where I should mark it? Thanks!

In provider's CHANGELOG.rst (see comment at the top of it).

But is it really a breaking change for the users of the Hook/Operator?

**Comment by potiuk** [comment]
Looked at it - there are many changes and it does look like breaking looking at the few of those. In CHANGELOG you should describe what the users of previous version should do to migrate to this one.

File: airflow/providers/weaviate/provider.yaml
Comment (eladkal) [inline airflow/providers/weaviate/provider.yaml:1]: Lets also have
```
versions:
  - 2.0.0
  ```
  
  Since this one is a breaking change

Comment (Lee-W) [reply airflow/providers/weaviate/provider.yaml:1]: Sure! just added

**Review by potiuk — APPROVED** [review]
Nice migration guide! Thanks @Lee-W for such detailed description.

**Comment by Lee-W** [comment]
> Nice migration guide! Thanks @Lee-W for such detailed description.

Thanks for providing your guidance. 🙌 🙏 I feel it would be great if we could have one or two more eyes on this one. So I'll keep it for some time.

File: airflow/providers/weaviate/provider.yaml
Comment (pankajastro) [reply airflow/providers/weaviate/provider.yaml:1]: I think we need to add 2.0.0 in the file provider.yaml too

**Review by pankajastro — APPROVED** [review]
lgtm

File: airflow/providers/weaviate/provider.yaml
Comment (Lee-W) [reply airflow/providers/weaviate/provider.yaml:1]: just pushed. didn't notice i've not yet pushed this change. 🤦‍♂️

**Pull request merged**